cabal-install/src/Distribution/Client/ProjectPlanning.hs

```diff
@@ -2856,26 +2868,105 @@ setRootTargets targetAction perPkgTargetsMap =
 --
 pruneInstallPlanPass1 :: [ElaboratedPlanPackage]
                       -> [ElaboratedPlanPackage]
-pruneInstallPlanPass1 pkgs =
-    map (mapConfiguredPackage fromPrunedPackage)
-        (fromMaybe [] $ Graph.closure graph roots)
+pruneInstallPlanPass1 pkgs
+    -- if there are repl target, we need to do a bit more work
+    | anyReplTarget = final_final_graph
+
+    -- otherwise we'll do less
+    | otherwise     = pruned_packages
   where
+    pkgs' :: [InstallPlan.GenericPlanPackage IPI.InstalledPackageInfo PrunedPackage]
     pkgs' = map (mapConfiguredPackage prune) pkgs
-    graph = Graph.fromDistinctList pkgs'
-    roots = mapMaybe find_root pkgs'
 
+    prune :: ElaboratedConfiguredPackage -> PrunedPackage
     prune elab = PrunedPackage elab' (pruneOptionalDependencies elab')
       where elab' =
                 setDocumentation
               $ addOptionalStanzas elab
 
+    graph = Graph.fromDistinctList pkgs'
+
+    roots :: [UnitId]
+    roots = mapMaybe find_root pkgs'
+
+    -- Make a closed graph by calculating the closure from the roots
+    pruned_packages :: [ElaboratedPlanPackage]
+    pruned_packages =  map (mapConfiguredPackage fromPrunedPackage) (fromMaybe [] $ Graph.closure graph roots)
+
+    closed_graph :: Graph.Graph ElaboratedPlanPackage
+    closed_graph = Graph.fromDistinctList pruned_packages
+
+    -- whether any package has repl targets enabled.
+    anyReplTarget :: Bool
+    anyReplTarget = any is_repl_gpp pkgs' where
+      is_repl_gpp (InstallPlan.Configured pkg) = is_repl_pp pkg
+      is_repl_gpp _                            = False
+
+      is_repl_pp (PrunedPackage elab _) = not (null (elabReplTarget elab))
+
+    -- Anything which is inplace and left after pruning could be a repl target, then just need to check the
+    -- reverse closure after calculating roots to capture dependencies which are on the path between roots.
+    -- In order to start a multi-repl session with all the desired targets we need to load all these components into
+    -- the repl at once to satisfy the closure property.
+    all_desired_repl_targets = [cp | InstallPlan.Configured cp <- fromMaybe [] $ Graph.revClosure closed_graph roots]
+
+    add_repl_target :: ElaboratedConfiguredPackage -> ElaboratedConfiguredPackage
+    add_repl_target ecp | ecp `elem` all_desired_repl_targets = ecp { elabReplTarget = maybeToList (ComponentTarget <$> (elabComponentName ecp) <*> pure WholeComponent)
+                                                                    , elabBuildStyle = BuildInplaceOnly InMemory }
+                        | otherwise = ecp
```

Comment:
In my personal taste, I prefer this:
```suggestion
    add_repl_target ecp 
      | ecp `elem` all_desired_repl_targets = 
          ecp { elabReplTarget = maybeToList (ComponentTarget <$> (elabComponentName ecp) <*> pure WholeComponent)
              , elabBuildStyle = BuildInplaceOnly InMemory }
      | otherwise = ecp
```
this comment is not a blocker and can be dismissed